### PR TITLE
Support for traits like behaviour

### PR DESCRIPTION
**Pull request by chetanism**
o Added support to define factory attributes via function
o Added support to pass on build options to attrs, build, create APIs. Note that build options is not yet supported by buildMany or createMany APIs.
o Refactored tests to move models, factories, adapters out of the main test files.
o Removed a couple of tests that checked afterCreate gets called, the following tests anyway test that. This got us rid of sinon spy stuff.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
 /* global window, define */
-(function(factory) {
+(function (factory) {
```
Comment (aexmachina) [inline index.js]: @chetanism thanks for the PR! You've introduced a lot of noise into this PR by changing the coding style. There's things I don't like about this coding style, but I wouldn't go ahead and change it in a PR without speaking to the author first.

File: index.js [hunk 2]
```diff
@@ -14,15 +14,15 @@
...
+  var Factory = function () {
     var factory = this,
-        factories = {},
-        defaultAdapter = new Adapter(),
-        adapters = {},
-        created = [];
+      factories = {},
+      defaultAdapter = new Adapter(),
+      adapters = {},
+      created = [];
```
Comment (aexmachina) [inline index.js]: Multi-line var statements should be indented with 4 spaces.

File: index.js (hunk 1 above)
Comment (aexmachina) [inline index.js]: There should be no space after `function` for anonymous functions.

File: index.js [hunk 3]
```diff
@@ -194,67 +199,89 @@
...
-              });
+              }
+              else {
+                fn.call(attrs, function (err, value) {
+                  if (err) return cb(err);
+                  attrs[key] = value;
+                  cb();
+                });
+              }
+            } else if(typeof fn === 'object') {
```
Comment (aexmachina) [inline index.js]: Space after `if` please :)

File: index.js [hunk 4]
```diff
@@ -194,67 +199,89 @@
...
-        }, function(err) {
-          if (err) return callback(err);
+          }, function (err) {
+            if (err) return cb1(err);
 
-          callback(null, attrs);
-        });
+            cb1(null, attrs);
+          });
+        }
```
Comment (aexmachina) [inline index.js]: Nice 👍

File: index.js [hunk 5]
```diff
@@ -371,31 +398,32 @@
     };
   };
 
-  var Adapter = function() {};
+  var Adapter = function () {
+  };
```
Comment (aexmachina) [inline index.js]: Prefer single line for empty blocks.

File: index.js [hunk 6]
```diff
@@ -408,18 +436,21 @@
     }
     return newObj;
   }
+
   function keys(obj) {
     var rv = [], key;
     for (key in obj) {
       if (obj.hasOwnProperty(key)) {
-       rv.push(key);
+        rv.push(key);
```
Comment (aexmachina) [inline index.js]: Thanks!

**Comment by aexmachina** [comment]
Thanks @chetanism, nice PR (except for the style changes!) I really this improvement. I think we do need to add this to `buildMany` and `createMany` as well, and the docs need to be updated too. Are you interested in doing these things?

**Comment by aexmachina** [comment]
I'm interested in adding another collaborator on this module. I suspect you'd be perfect, are you interested?

**Comment by aexmachina** [comment]
Also be aware that you'll need to merge from master because #50 has been merged.

**Comment by chetanism** [comment]
@aexmachina, sorry for breaking the style rules, it's  just my IDE set to fix styles mostly as per the [javascript style guide](https://github.com/airbnb/javascript).

I understand your point, and am ready to revert style changes as well, but still I would like to propose to keep the style as per the above mentioned style guide. Going ahead it would help to follow a default standard definition for other contributors as well.

As for being a collaborator, I am definitely interested to be part of this project!

**Comment by chetanism** [comment]
In the mean time, I'll merge #50, to my changes.

**Comment by aexmachina** [comment]
Yeah I understand. I don't like the style either. Happy to change it but it doesn't belong in this PR. No need to change it back, I'm more providing feedback on good PR etiquette. 

How can I contact you? My contact details are on my profile?

**Comment by chetanism** [comment]
Agreed! Will keep in mind to not pollute PRs with such stuff. Instead will take these things separately.

**Pull request merged**